src/api/src/middleware/authenticate.ts

```diff
@@ -8,40 +8,43 @@ import jwt from 'jsonwebtoken';
 import { kSQL } from '../tokens';
 import { discordOAuth2 } from '../util';
 
+interface TokenData {
+	sub: string;
+}
+
+export interface OAuthInfo {
+	token: string;
+	userId: string;
+}
+
 export default async (req: Request, _: Response, next?: NextHandler) => {
 	if (req.headers.cookie) {
 		const sql = container.resolve<postgres.Sql<any>>(kSQL);
 		const cookies = cookie.parse(req.headers.cookie) as { token: string };
 
-		let decoded: { provider: 'Discord' | 'Twitch'; access_token: string };
+		let decoded: TokenData;
 		try {
-			decoded = jwt.verify(cookies.token, process.env.JWT_SECRET!) as {
-				provider: 'Discord' | 'Twitch';
-				access_token: string;
-			};
+			decoded = jwt.verify(cookies.token, process.env.JWT_SECRET!) as TokenData;
 
-			req.oauth = { provider: decoded.provider, token: cookies.token, access_token: decoded.access_token };
+			req.oauth = { token: cookies.token, userId: decoded.sub };
 			return next?.();
 		} catch (error) {
 			if (error.name === 'TokenExpiredError') {
-				const { provider, access_token } = jwt.decode(cookies.token) as {
-					provider: 'Discord' | 'Twitch';
-					access_token: string;
-				};
-				const [connection] = await sql<{ user_id: string; refresh_token: string }>`
-					select user_id, refresh_token
+				const { sub } = jwt.decode(cookies.token) as TokenData;
+				const [connection] = await sql<{ refresh_token: string }>`
+					select refresh_token
 					from connections
-					where access_token = ${access_token}
-						and provider = ${provider};`;
+					where user_id = ${sub}
+						and provider = "Discord";`;
```

Comment:
We can't hardcode the provider in the middleware, putting it inside the JWT should be no issue.

This is to determine with which service the user logged in because both discord and twitch have different restrictions (for example Twitch requires a middleware that calls their `/validate` endpoint every time before you use their token)

Reply:
I don't think we need to issue tokens depending on which service the user has logged in with. We have access to all of their tokens in the database, so only storing their user ID in the token is enough for us to know the credentials for every connected service. We don't need to validate their access tokens every time they make an API call, only when they attempt to make an API call that requires a connected service.

I agree that hardcoding the provider here isn't great, so I'll work on a solution for that. However, I do not think we should issue provider-specific tokens.

Reply:
Sure, but then we might as well remove providers all together from the `select` queries and tokens and keep it just in the `insert` queries and database.